### PR TITLE
Add ssl support

### PR DESCRIPTION
**Pull request by purbon**
Incorporate #1, rebase it to update the code include the recent  and add test to it.

### CODE REVIEW

**Comment by purbon** [comment]
@jsvd would you be to nice to review this, so we could move forward and merge! thanks man!

File: lib/logstash/inputs/relp.rb [hunk 1]
```diff
@@ -87,12 +131,24 @@ def run(output_queue)
...
+                      :backtrace => ssle.backtrace)
       rescue LogStash::ShutdownSignal
         @interrupted = true
         break
+      rescue IOError, Interrupted
+        if @interrupted
+          # Intended shutdown, get out of the loop
+          @relp_server.shutdown
+          break
+        else
```
Comment (jsvd) [inline lib/logstash/inputs/relp.rb]: doing else + raise is the same as not rescuing at all, right?

File: lib/logstash/inputs/relp.rb [hunk 2]
```diff
@@ -27,6 +27,25 @@ class LogStash::Inputs::Relp < LogStash::Inputs::Base
   # The port to listen on.
   config :port, :validate => :number, :required => true
 
+  # Enable SSL (must be set for other `ssl_` options to take effect).
+  config :ssl_enable, :validate => :boolean, :default => false
+
+  # Verify the identity of the other end of the SSL connection against the CA.
+  # For input, sets the field `sslsubject` to that of the client certificate.
+  config :ssl_verify, :validate => :boolean, :default => false
```
Comment (jsvd) [inline lib/logstash/inputs/relp.rb]: this should be on by default, and print a huge warning if set to false.
If a person is setting up ssl then turning off validation defeats the purpose..well..red alert..

More on this in the [logstash-output-elasticsearch](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch.rb#L279-L282)

Comment (jsvd) [inline lib/logstash/inputs/relp.rb]: and [here](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch.rb#L646-L652)

File: lib/logstash/inputs/relp.rb [hunk 3]
```diff
@@ -27,6 +27,25 @@ class LogStash::Inputs::Relp < LogStash::Inputs::Base
...
+  config :ssl_cacert, :validate => :path
+
+  # SSL certificate path
+  config :ssl_cert, :validate => :path
+
+  # SSL key path
+  config :ssl_key, :validate => :path
+
+  # SSL key passphrase
+  config :ssl_key_passphrase, :validate => :password, :default => nil
```
Comment (jsvd) [inline lib/logstash/inputs/relp.rb]: I don't think a default is necessary here (if it's being set to nil)

File: lib/logstash/inputs/relp.rb [hunk 4]
```diff
@@ -36,16 +55,41 @@ def initialize(*args)
...
+    @relp_server = RelpServer.new(@host, @port,['syslog'], @ssl_context)
   end # def register
 
   private
+  def initialize_ssl_context
+    require "openssl"
+
+    @ssl_context = OpenSSL::SSL::SSLContext.new
+    @ssl_context.cert = OpenSSL::X509::Certificate.new(File.read(@ssl_cert))
+    @ssl_context.key = OpenSSL::PKey::RSA.new(File.read(@ssl_key),@ssl_key_passphrase)
```
Comment (jsvd) [inline lib/logstash/inputs/relp.rb]: if `ssl_key_passphrase` is a `:password` then you need to user the .value method to actually get the password content from that object

File: lib/logstash/inputs/relp.rb (hunk 2 above)
Comment (purbon) [inline lib/logstash/inputs/relp.rb]: Ok this, makes sense.

File: lib/logstash/inputs/relp.rb (hunk 4 above)
Comment (purbon) [inline lib/logstash/inputs/relp.rb]: cool, thanks. was not aware of it.

File: lib/logstash/inputs/relp.rb (hunk 3 above)
Comment (purbon) [inline lib/logstash/inputs/relp.rb]: If I remove the nill default, test are handing, so I guess some default is required here.

File: lib/logstash/inputs/relp.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/inputs/relp.rb]: yeah, I think the idea here is to catch Interrupt and IOError that might happen when the interrupt happen, but have no more idea of how this relp thing works in real life, so can not say if this catch, raise makes sense.

**Comment by purbon** [comment]
@jsvd done with most of your chances, thanks for your time man!.

File: lib/logstash/util/relp.rb [hunk 5]
```diff
@@ -118,7 +118,10 @@ def initialize(host,port,required_commands=[])
                     :host => host, :port => port)
       raise
     end
-    @logger.info? and @logger.info("Started RELP Server", :host => host, :port => port)
+    if ssl_context
+      @server = OpenSSL::SSL::SSLServer.new(@server, ssl_context)
+    end
+    @logger.info? and @logger.info("Started #{ssl_context ? 'SSL-enabled ' : ''}RELP Server", :host => host, :port => port)
```
Comment (jsvd) [inline lib/logstash/util/relp.rb]: since this a call that happens during initalize then it's not worth having the .info? and .info performance pattern

**Comment by purbon** [comment]
@jsvd done with your last request, let me know if this is good to proceed, or you still have open questions. Thank you for your time.

**Comment by jsvd** [comment]
LGTM
Note: I didn't do a real life test, only ran the spec suite

**Comment by purbon** [comment]
@jsvd yeah! me too ... any idea how to run a real life test for it?

**Pull request closed without merging**